Remove all mentions of "data files" and "garbage collection" remaining. (#884)
GothenburgBitFactory/tw.org: content/docs/ids.md

# ID Numbers

Taskwarrior keeps the current tasks (mostly pending and recurring tasks) in the "working set," and assigns ID numbers to them.
The ID number is the index (1-based) of the tasks in the working set.
When a task is completed or deleted, it is moved out of the working set, and loses the ID.
All tasks, in the working set or otherwise, have a UUID, and are always addressable by UUID.

By having a working set, average performance is improved.
This works because most of the commands and reports are based on only the pending tasks, which are in the working set.
The working set tends to stabilize at a constant size, regardless of the amount of work you are getting done [1].
The total number of tasks, on the other hand, grows unbounded.
This means that some reports, such as the `completed` or `all` report will run slower over time because there is more data to read and display.

One nice side effect of this mechanism is that the ID numbers of pending tasks remain small, and less prone to error during data entry.

One downside is that ID numbers change occasionally, although in a predictable way.

> ## Rebuilding

Reviewer: ```suggestion
## Rebuilding the Working Set
```

When a task is marked as deleted, it gets a new status, and an `end` attribute, among other things.
The task is still in the working set, but it doesn't belong there.
There is an operation called "rebuilding" that is automatically run by taskwarrior to move tasks out of the working set.
When moving tasks out of the working set, the ID numbers are changed, because they should be a sequence with no gaps assigned to all tasks in the working set.

> But rebuilding is only run occasionally: it is run immediately before any report that displays ID numbers, but not when a task is modified in any way.

Reviewer: ```suggestion
But the working set is only rebuilt occasionally: immediately before any report that displays ID numbers, but not when a task is modified in any way.
```

The last command shows ID numbers again.
Clearly, while the middle three commands are being run, it would be a mistake to renumber any tasks.
When the last command is run, new ID numbers are shown, and any subsequent commands would use those.
This is why only reports that display ID numbers perform a rebuild first.

## Configuration

> The `gc` configuration setting may be set to 'off' to disable the rebuild operation.

Reviewer: I'm curious what your thoughts are here? The config uses `gc`, but we're trying to use "rebuild" everywhere else..

Author: I would prefer to rename the config from `gc` to something like `rebuildworkingset`. But this depends on how you want to handle backwards compatibility.

Reviewer: Let's leave this as-is, and please open a bug in GothenburgBitFactory/taskwarrior to rename the config option. We can keep the old one for backward compatibility.

Author: Sure! https://github.com/GothenburgBitFactory/taskwarrior/issues/3557
